File: ios/Sources/App/Services/GitHubService.swift
import Foundation

struct GitHubUser: Decodable {
    let login: String
}

struct GitHubIssue: Decodable, Identifiable {
    let id: Int
    let number: Int
    let title: String
    let state: String
    let htmlURL: URL
    let user: GitHubUser
    let pullRequest: GitHubPullRequestMarker?

    enum CodingKeys: String, CodingKey {
        case id, number, title, state, user
        case htmlURL = "html_url"
        case pullRequest = "pull_request"
    }
}

struct GitHubPullRequestMarker: Decodable {}

struct GitHubPullRequest: Decodable, Identifiable {
    let id: Int
    let number: Int
    let title: String
    let state: String
    let draft: Bool
    let htmlURL: URL
    let user: GitHubUser

    enum CodingKeys: String, CodingKey {
        case id, number, title, state, draft, user
        case htmlURL = "html_url"
    }
}

enum GitHubServiceError: Error, LocalizedError {
    case missingToken
    case invalidResponse
    case api(Int, String)

    var errorDescription: String? {
        switch self {
        case .missingToken: return "Add a GitHub personal access token in Settings."
        case .invalidResponse: return "GitHub returned an invalid response."
        case .api(let status, let message): return "GitHub error \(status): \(message)"
        }
    }
}

actor GitHubService {
    static let shared = GitHubService()

    func validateToken() async throws -> GitHubUser {
        try await request(path: "/user")
    }

    func issues(repository: String) async throws -> [GitHubIssue] {
        let values: [GitHubIssue] = try await request(path: "/repos/\(repository)/issues?state=open&per_page=30")
        return values.filter { $0.pullRequest == nil }
    }

    func pullRequests(repository: String) async throws -> [GitHubPullRequest] {
        try await request(path: "/repos/\(repository)/pulls?state=open&per_page=30")
    }

    private func request<Value: Decodable>(path: String) async throws -> Value {
        let token: String
        do { token = try KeychainService.githubToken() } catch { throw GitHubServiceError.missingToken }
        guard let url = URL(string: "https://api.github.com\(path)") else { throw GitHubServiceError.invalidResponse }
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        request.setValue("2026-03-10", forHTTPHeaderField: "X-GitHub-Api-Version")
        request.setValue("TermVault-iOS", forHTTPHeaderField: "User-Agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw GitHubServiceError.invalidResponse }
        guard 200..<300 ~= http.statusCode else {
            let message = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String ?? "Request failed"
            throw GitHubServiceError.api(http.statusCode, message)
        }
        return try JSONDecoder().decode(Value.self, from: data)
    }
}

File: android/app/src/main/kotlin/dev/termvault/app/github/GitHubService.kt
package dev.termvault.app.github

import dev.termvault.app.security.SecretStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory

sealed class GitHubServiceError(message: String) : Exception(message) {
    object MissingToken : GitHubServiceError("Add a GitHub personal access token in Settings.")
    class Api(status: Int, apiMessage: String) : GitHubServiceError("GitHub error $status: $apiMessage")
}

/** Mirrors iOS `GitHubService.swift` — validate a PAT, list open issues/PRs for a `owner/repo` slug. */
class GitHubService(private val secretStore: SecretStore) {
    private val json = Json { ignoreUnknownKeys = true }

    private val client = OkHttpClient.Builder()
        .addInterceptor(Interceptor { chain ->
            val request = chain.request().newBuilder()
                .header("Accept", "application/vnd.github+json")
                .header("X-GitHub-Api-Version", "2026-03-10")
                .header("User-Agent", "TermVault-Android")
                .build()
            chain.proceed(request)
        })
        .build()

    private val api: GitHubApi = Retrofit.Builder()
        .baseUrl("https://api.github.com/")
        .client(client)
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .build()
        .create(GitHubApi::class.java)

    suspend fun validateToken(): GitHubUser = withContext(Dispatchers.IO) {
        perform { api.user(bearerToken()) }
    }

    suspend fun issues(repository: String): List<GitHubIssue> = withContext(Dispatchers.IO) {
        val (owner, repo) = splitRepository(repository)
        perform { api.issues(owner, repo, bearerToken()) }.filter { it.pullRequest == null }
    }

    suspend fun pullRequests(repository: String): List<GitHubPullRequest> = withContext(Dispatchers.IO) {
        val (owner, repo) = splitRepository(repository)
        perform { api.pullRequests(owner, repo, bearerToken()) }
    }

    private fun splitRepository(repository: String): Pair<String, String> {
        val parts = repository.trim().trimStart('/').split("/")
        require(parts.size == 2) { "Repository must be in \"owner/repo\" form." }
        return parts[0] to parts[1]
    }

    private fun bearerToken(): String {
        val token = secretStore.githubToken() ?: throw GitHubServiceError.MissingToken
        return "Bearer $token"
    }

    private suspend fun <T> perform(call: suspend () -> T): T {
        try {
            return call()
        } catch (e: HttpException) {
            val body = e.response()?.errorBody()?.string()
            val message = body?.let { runCatching { json.parseToJsonElement(it).jsonObject["message"]?.jsonPrimitive?.content }.getOrNull() }
            throw GitHubServiceError.Api(e.code(), message ?: "Request failed")
        }
    }
}
